OpenSim/Moco/MocoFrameOrientationConstraint.cpp

```diff
@@ -0,0 +1,127 @@
+
+MocoFrameOrientationConstraint::MocoFrameOrientationConstraint() {
+    constructProperties();
+}
+void MocoFrameOrientationConstraint::initializeOnModelImpl(const Model& model, 
+        const MocoProblemInfo&) const {
+
+    int nFramePairs = getProperty_frame_pairs().size();
+    MocoConstraintInfo info;
+    std::vector<MocoBounds> bounds;
+
+    for (int i = 0; i < nFramePairs; ++i) {
+        const auto frame1_path = get_frame_pairs(i).get_frame1_path();
+        OPENSIM_THROW_IF(!model.hasComponent<Frame>(frame1_path), Exception,
+            "Could not find frame '{}'.", frame1_path);
+        auto& frame1 = model.getComponent<Frame>(frame1_path);
+        const auto frame2_path = get_frame_pairs(i).get_frame2_path();
+        OPENSIM_THROW_IF(!model.hasComponent<Frame>(frame2_path), Exception,
+            "Could not find frame '{}'.", frame2_path);
+        auto& frame2 = model.getComponent<Frame>(frame2_path);
+        m_frame_pairs.emplace_back(&frame1, &frame2);
+
+        const double& minimum = get_frame_pairs(i).get_minimum_angle();
+        const double& maximum = get_frame_pairs(i).get_maximum_angle();
+        
+        OPENSIM_THROW_IF(minimum > maximum, Exception,
+            "Expected the minimum angle for this frame pair "
+            "to be less than or equal to the maximum angle, "
+            "but they are {} and {}, respectively.",
+            minimum, maximum);
+        bounds.emplace_back(minimum, maximum);
+    }
+    m_coordinateAxis = get_coordinate_axis();
+    OPENSIM_THROW_IF(((m_coordinateAxis!=0) && (m_coordinateAxis != 1) && 
+        (m_coordinateAxis != 2) && (m_coordinateAxis != 3)), Exception,
+        "The coordinate axis should be 0,1,2,or 3.", m_coordinateAxis);
+
+    setNumEquations(nFramePairs);
+    info.setBounds(bounds);
+    const_cast<MocoFrameOrientationConstraint*>(this)->setConstraintInfo(info);
+
+}
+
+void MocoFrameOrientationConstraint::calcPathConstraintErrorsImpl(
+        const SimTK::State& state, SimTK::Vector& errors) const {
+    int iconstr = 0;
+
+    getModel().realizePosition(state);
+
+    for (const auto& frame_pair : m_frame_pairs) {
+        const auto& frame1_rotation = frame_pair.first->getRotationInGround(
+                state);
+        const auto& frame2_rotation = frame_pair.second->getRotationInGround(
+                state);
+        
+        if (m_coordinateAxis == 0 || m_coordinateAxis == 1 ||
+                m_coordinateAxis == 2) {
+            
+            auto frame1_angle = 
+                frame1_rotation.convertOneAxisRotationToOneAngle(
+                        SimTK::CoordinateAxis(m_coordinateAxis));
+            auto frame2_angle = 
+                    frame2_rotation.convertOneAxisRotationToOneAngle(
+                            SimTK::CoordinateAxis(m_coordinateAxis));
+            auto relative_angle = frame2_angle - frame1_angle;
+            errors[iconstr++] = relative_angle;
+        }
+        else if(m_coordinateAxis == 3) {

```

Comment:
This will now be the top level if-statement where you just check the boolean vector and the other branch can be handled with `else` (so you can avoid the cumbersome or-statements).